cbmc/proofs/mqtt_state.c

```diff
@@ -0,0 +1,556 @@
+
+  return
+#ifdef TOPIC_LENGTH_MAX
+    pElt->topicFilterLength < TOPIC_LENGTH_MAX &&
+#endif
+    pElt->topicFilterLength < CBMC_MAX_OBJECT_SIZE - sizeof( _mqttSubscription_t ) &&
+    __CPROVER_r_ok(pElt->pTopicFilter, pElt->topicFilterLength) &&
+    __CPROVER_w_ok(pElt->pTopicFilter, pElt->topicFilterLength);
+}
+
+// Subscription linked lists
+
+IotListDouble_t *allocate_IotMqttSubscriptionList(IotListDouble_t *pSub,
+						  size_t length)
+{
+  if ( pSub == NULL ) pSub = malloc_can_fail( sizeof( *pSub ) );
+  if ( pSub == NULL ) return NULL;
+
+  IotListDouble_Create(pSub);
+  for (int i = 0; i < length; i++) {
+    _mqttSubscription_t *pElt = allocate_IotMqttSubscriptionListElt(NULL);
+    __CPROVER_assume(pElt);
+    IotListDouble_InsertHead( pSub, &(pElt->link) );
+  }
+  return pSub;
+}
+
+bool valid_IotMqttSubscriptionList(IotListDouble_t *pSub, size_t length)
+{
+  if ( pSub == NULL ) return 0;
+
+  bool result = 1;
+  IotListDouble_t *pLink;
+  IotContainers_ForEach( pSub, pLink ) {
+    _mqttSubscription_t
+      *pElt = IotLink_Container( _mqttSubscription_t, pLink, link );
+    result = result && valid_IotMqttSubscriptionListElt( pElt );
+  }
+
+  return
+#ifdef SUBSCRIPTION_COUNT_MAX
+    length < SUBSCRIPTION_COUNT_MAX &&
+#endif
+    result;
+}
+
+/****************************************************************
+ * IotMqttPublishInfo
+ ****************************************************************/
+
+IotMqttPublishInfo_t *allocate_IotMqttPublishInfo(IotMqttPublishInfo_t *pInfo)
+{
+  if ( pInfo == NULL ) pInfo = malloc_can_fail( sizeof( *pInfo ) );
+  if ( pInfo == NULL ) return NULL;
+
+  pInfo->pTopicName = malloc_can_fail( pInfo->topicNameLength );
+  // assumption here is checked below in valid_
+  __CPROVER_assume( VALID_CBMC_SIZE( pInfo->payloadLength ) );
+  pInfo->pPayload = malloc_can_fail( pInfo->payloadLength );
+  return pInfo;
+}
+
+bool valid_IotMqttPublishInfo(IotMqttPublishInfo_t *pInfo)
+{
+  return
+    pInfo &&
+
+    VALID_STRING(pInfo->pTopicName, pInfo->topicNameLength) &&
+    VALID_CBMC_SIZE( pInfo->topicNameLength ) &&
+    pInfo->pTopicName != NULL &&
+
+    VALID_STRING(pInfo->pPayload, pInfo->payloadLength) &&
+    VALID_CBMC_SIZE( pInfo->payloadLength ) &&
+
+    pInfo->retryMs <= IOT_MQTT_RETRY_MS_CEILING &&
+    pInfo->retryMs > 0  &&
+
+    // BUG: try without this name length assumption
+    pInfo->topicNameLength < 0xFFFF &&
+    // BUG: try without this name length assumption
+    pInfo->payloadLength > 0;
+}
+
+/****************************************************************
+ * IotNetworkConnection
+ ****************************************************************/
+
+void *allocate_IotNetworkConnection()
+{
+  return allocate_opaque_type();
+}
+
+/****************************************************************
+ * IotNetworkInterface
+ ****************************************************************/
+
+IotNetworkInterface_t *allocate_IotNetworkInterface()
```

Comment:
Why does this one not take an input parameter?

Reply:
This is just one type that appears everywhere as a pointer to a struct (and not a substruct itself), so there is no need for the complexity here.